Sources/Common/Background Queue/QueueRunRequest.swift

```diff
@@ -38,27 +38,35 @@ public class CioQueueRunRequest: QueueRunRequest {
     }
 
     private func startNewRequestRun() {
-        let inventory = storage.getInventory()
-
-        runTasks(queueInventory: inventory, queryTotalNumberTasks: inventory.count)
+        runTasks(lastRanTask: nil)
     }
 
     // Disable swiftlint because function at this time isn't too complex to need to make it smaller.
     // Many of the lines of this function are logging related.
     // swiftlint:disable:next function_body_length
     private func runTasks(
-        queueInventory: [QueueTaskMetadata],
-        queryTotalNumberTasks: Int,
+        lastRanTask: QueueTaskMetadata?,
         lastFailedTask: QueueTaskMetadata? = nil
     ) {
-        guard let nextTaskToRunInventoryItem = queryRunner.getNextTask(queueInventory, lastFailedTask: lastFailedTask)
+        // get the inventory before running each task. If a task was added to the queue while the last task was being
+        // executed, we can assert that new task will execute during this run.
+        let queueInventory = storage.getInventory()
```

Comment:
This is the major piece of the BQ refactor. Instead of iterating a snapshot of the queue tasks inventory, we re-query all of the queue inventory tasks before running each task. That way if new tasks get added to the queue while running the queue, they get executed this BQ run instead of waiting until a future run.

Reply:
So would it be justified to say that if a BG starts executing the tasks, it will keep running until it reaches end of queue, no matter if it had few number of tasks at the time of start ?

Reply:
> no matter if it had few number of tasks at the time of start 

The queue will keep running until it reaches the end of the queue, even if more tasks *are added* while the queue is running. Tasks do not get removed from the queue so there isn't a scenario where there would be fewer tasks then at the time of start.